### ISSUE TITLE
Support visualisation of tracks on map (GPX-format)

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#2469]
It would be a valuable feature if it could be possible to load GPX-files containing tracking points `<trkpt>` in order to show this track on the (live)map.
On some caching trails or caching hiking tours such GPX-tracks are provided to the cacher to be able to follow the route (to the cache).
Especially in countries with bad map coverage (in terms of hiking trails) like e.g. Greece this seems to be a helper which is used quite often.

---

It could be implemented to the live map menu "Map modes"->"Load track" (and "Show track"/"Hide track") like already implemented for the position trail.

**Comment by andi84prv** [comment]
It’s exactly the function which I’m looking for:
I like to import tracks from gpx-files and there is no need to record anything. My use case would be:
Import track from a hiking tour for following the right ways and searching some geocaches at the same time. Now I have to switch between two different apps (one for showing the gpx-track and cgeo for caching).
For me it would be enough, when there would be a possibility for showing the track on the live map. So it would be possible to follow the track and looking for some geocaches in the near area. It really would be amazing for me!

**Comment by Lineflyer** [comment]
@andi84prv 
Given that this is not yet implemented and not planned in near future I would like to suggest a workaround for you: Do you by chance use a Samsung device?
Because as Samsung devices support multiwindow, I do use a split screen of Locus and c:geo as a workaround..

**Comment by andi84prv** [comment]
@Lineflyer 
Actually I use a Samsung device. More precisely it's the SM-G800F with Andorid 4.4.2.
As I can see there are some possibilities in the appstore. Would you recommend one? Or just "try and error"?
Thanks for the hint! It sound absolutly useful although it's only a workaround!! Anyway the best solution would be the possibility to import gpx-tracks in c:geo (let's just dream about it ;).

**Comment by Lineflyer** [comment]
@andi84prv 
The Multiwindow feature should be on-board, no need for a dedicated app.
I do use the app "Locus" for all track related activities (the free version should do the job although I use the full version).

**Comment by moving-bits** [comment]
Just to give you a short sneak preview. I have a first WIP for

- parsing GPX track files
- importing tracks to memory
- displaying a track on NewMap
- hiding/unhiding tracks

Next step is to adapt this to Google Maps as well...

Loading a track file from map menu entry "Load track":
![image](https://user-images.githubusercontent.com/3754370/82749350-f1803800-9da8-11ea-976e-2232280db1eb.png)

Displaying a track:
![image](https://user-images.githubusercontent.com/3754370/82749330-c695e400-9da8-11ea-9587-f275078fa9ce.png)

Hide/unhide track display using the map's hide menu:
![image](https://user-images.githubusercontent.com/3754370/82749337-ceee1f00-9da8-11ea-9399-9eb9170e61fc.png)

**Comment by dcjkfgdjhd** [comment]
Looks awesome so far! Really looking forward to it!

Here is a tiny idea that would be convenient: When importing caches from GPX, if that GPX _also_ contains a track, that track should automatically be 'loaded' into the map as well. 

I guess it would be good if that were optional; it could be a checkbox within the GPX import dialogue. But _probably_ in 95% of the cases, when people want to import a track into c:geo, it is because they planned an entire  caching tour ahead of time, so the situation where the GPX contains _both_ caches and a track is probably the typical case, so most of the time you probably _do_ want to also load the track when importing caches.

**Comment by alan666notb** [comment]
Would it also work with gpx files containing routes, rather than tracks.
I know the format is different, don't know how.

**Comment by moving-bits** [comment]
> Would it also work with gpx files containing routes, rather than tracks.
> I know the format is different, don't know how.

Should be technically possible, but if I understand [this Wikipedia article](https://en.wikipedia.org/wiki/GPS_Exchange_Format) right, there would be an additional step of calculating a route from each route point to the next.

**Comment by Lineflyer** [comment]
The first approach looks quite good IMHO.
Not sure yet about the menu entries: 
From what I understand "Hide track" will hide it on the map but it is still "selected"/"loaded"?
If yes, that's fine, but do we also need an action to "unload" it completely?

Technical question (just for my background knowledge):
Is this tracked stored somewhere inside c:geo after loading it (e.g. in our database) or only loaded from the file and hold in memory while c:geo is running?

**Comment by moving-bits** [comment]
Thanks for your feedback.

> The first approach looks quite good IMHO.
> Not sure yet about the menu entries:
> From what I understand "Hide track" will hide it on the map but it is still "selected"/"loaded"?

yes

> If yes, that's fine, but do we also need an action to "unload" it completely?

yes - I thought about a "clear track" menu entry, grouping the three "clear xxx" entries into its own submenu. (Not yet implemented.)

> Technical question (just for my background knowledge):
> Is this tracked stored somewhere inside c:geo after loading it (e.g. in our database) or only loaded from the file and hold in memory while c:geo is running?

Nothing is stored to c:geo so far except "hide track" status and "track filename".
If "track filename" is set, the file gets reloaded automatically upon opening/resuming the map.

One side note: Currently I'm displaying the first track only (for memory and performance reasons). A GPX file can hold multiple tracks. Maybe some sort of selector could be added later, or multiple tracks could be displayed.

**Comment by dcjkfgdjhd** [comment]
> there would be an additional step of calculating a route from each route point to the next.

That is correct.

**Comment by Lineflyer** [comment]
Thanks..
Lets take this then as first implementation and file the other ideas (which sound good) as further improvement issues.

Regarding the menus:
Together with recently added items the map menu grows. Let me revisit the whole thing and see if we find meaningful groups. This should not stop from merging the current state though.

One testing note:
Please handle (if not yet), that the file selected could be removed by the user meanwhile. We should IMHO in this case just remove the file reference and fail silently (assuming the user just does not need this track anymore).

**Comment by moving-bits** [comment]
> This should not stop from merging the current state though.

We should wait for the version after the next release, IMHO, to get the current work out on the street. It's already pretty much waiting on master to be released, and we are ready to start a beta phase, IMHO, except for the ongoing work on #8260.

**Comment by Lineflyer** [comment]
You are probably right :)

**Comment by seballaA** [comment]
I am very excited about this upcoming feature, I am craving for it for quite some time as my use of c:geo is 30% geocaching, and 70% maps for hiking and biking. I use c:geo for seven years now, about the period this issue is open. I am looking forward to this and I would like to thank the team for this great app with a donation which I am going to do : now.

**Comment by moving-bits** [comment]
The original request (visualize a track from GPX file) is implemented now with #8376, so let's close this old issue. If you have request for improvement or feature request, please open dedicated new issues, which makes it easier to track.

**Comment by MagpieFourtyTwo** [comment]
There seem to be some issues when displaying (resp trying to display) a track from a GPX file:

- after choosing an incorrectly formatted file (like the one exported with the previous track export version, with two `lat` but no `lon` tag) c:geo first stutters, then hangs, and finally crashes

- same happens when the previously chosen track file does not exist anymore (BTW: don't we need a "remove track from map" menu entry?)

Only solution: delete entry `trackfile` from System -> Show Settings. 
BTW: seems as even Map view -> Hide -> Track does not help in these scenarios.

For me it seems as if there is no plausibility check for the track data (valid `lat` and `lon`, at least) and no check for file existence ...? OTOH: If this is already implemented, I probably have another issue and can of course provide a log ...

**Comment by Lineflyer** [comment]
@MagpieFourtyTwo While commenting here makes sense to keep on the topic, it makes IMHO however more sense to open dedicated issues for each problem you find. This helps to keep track of open problems. So it would be nice if you could take those two valid problems into two new issues, ideall with step-by-step to reproduce.

Thanks for your effort in helping to test :)

**Comment by moving-bits** [comment]
> There seem to be some issues when displaying (resp trying to display) a track from a GPX file:

@MagpieFourtyTwo 
I've created #8480, #8481, #8482 for it with minimum text - feel free to add more info if needed.

**Comment by MagpieFourtyTwo** [comment]
Thx! Could not do it on my own, as I was out in the wild wild world, to perform some real life tests. ;)

**Comment by jakedot** [comment]
Hmmm, if we reuse the "Import GPX" menu item like suggested here https://github.com/cgeo/cgeo/issues/8020#issuecomment-632844750 then the tracks imported should be in c:geo permanently, just like the caches are imo. Maybe we could integrate them as a bogus cache type? Then you could have a menu item for that type to control a track being shown on the map or not. Optionally, give the track a "position" by it's starting point or the middle or whatever, and make it possible to toggle displaying tracks from the map too?

**Comment by moving-bits** [comment]
> Hmmm, if we reuse the "Import GPX" menu item like suggested here [#8020 (comment)](https://github.com/cgeo/cgeo/issues/8020#issuecomment-632844750) then the tracks imported should be in c:geo permanently, just like the caches are imo. Maybe we could integrate them as a bogus cache type? 

Sorry, but caches and tracks are completely different data structures. Both can be transported via a GPX file, but that's nearly the only thing they have in common. At a first quick glance we would need to stretch our concept of caches and waypoints quite a bit to press GPX tracks into it (and still would have to handle the drawing etc. separately). Don't know how c:geo would behave with "caches" having hundreds or even thousands of "waypoints"...

> Then you could have a menu item for that type to control a track being shown on the map or not.

There is an option to hide/unhide tracks in the map's "hide" menu.

**Comment by MagpieFourtyTwo** [comment]
My 2 Cent: Displaying tracks is an IMHO very special feature, which most of the c:geo users will not even think of, I guess (talked to some, and most of them did not even know, what a track is). So I would consider a major change in data structure as overdone.

I would suggest to basically stay with what we have: One track at a time seems (at least to me) way more than most of us will ever need.

Moreover GPX file im- and export is an extreme complex thing, and even Garmin, who deals with this topic for decades now, hasn't got it foolproof yet. There are too many different implementations and even more possibilities to interpret the GPX rules, so c:geo (whose main purpose is a completely different one anyway) should not even try to deal with this neverending story, as it will give more problems than real gain for cacher's daily life.

I would say there are some way more important (and success promising) topics to deal with.  ;)

**Comment by alan666notb** [comment]
I would use this feature.

When we go out to do a series, I often plot a route around it to follow (But note - 'route', not 'track' - I do it using the GME Chrome extension on the normal geocaching.com maps).

At the moment, I load the track into another mapping program (AlpineQuest) and view it on that app. It'd be really useful to see the route on the same map that shows the caches I'm looking for.

**Comment by MagpieFourtyTwo** [comment]
Yep, agreed - and I do it the same, esp. on long tours with a group.
But, how many tracks do you need to display at the same time ...?
And would it be really necessary to have them integrated in the c:geo database ...?

**Comment by MagpieFourtyTwo** [comment]
Whereas I mainly use the track export - to improve OSM. I use the export as template for drawing missing tracks resp. correcting improper drawn ones.

**Comment by alan666notb** [comment]
I would only need one route at a time.

Maybe all it needs is to define a GPX filename under the system->maps settings, if the file is there, display the route/track, otherwise, don't.

I can't imagine why I would want to load & display a track, unless I wanted to follow a track someone else had already done.

**Comment by Lineflyer** [comment]
I think what we have now is a good basic implementation of tracks. As c:geo is mainly no hiking app, but tracks are really useful while e.g. doing a hiking/biking cache, the current implementation is sufficient.

An extension could be to handle tracks like our offline maps: Set a directory and use a dropdown menu of all track files in that directory to pick one from them to be shown. If you agree, we could/should open another feature request for that feature extension.

Not to forget: @moving-bits Thanks for implementation!!

**Comment by Lineflyer** [comment]
> Whereas I mainly use the track export - to improve OSM. I use the export as template for drawing missing tracks resp. correcting improper drawn ones.

Offtopic: For this purpose I mostly use the corresponding Locus Maps feature / addin.

**Comment by MagpieFourtyTwo** [comment]
I so far used "Track My Trip" - but why have two apps running, if the best of all has the possibility anyway?  ;)

**Comment by moving-bits** [comment]
> An extension could be to handle tracks like our offline maps: Set a directory and use a dropdown menu of all track files in that directory to pick one from them to be shown. If you agree, we could/should open another feature request for that feature extension.

Although this sounds nice, I'm not sure if we really need this, as using the map menu entry "load track" opens c:geo's internal file selector which lists all GPX files in the last selected dir + allows for selecting a different dir.

**Comment by MagpieFourtyTwo** [comment]
Not really - different dir selection does not work over here ... and I think *this* was it, for which I wanted to open a new issue and could not remember anymore ...  ;)

**Comment by MagpieFourtyTwo** [comment]
After I have selected a different dir I can not select the GPX to be imported, as immediately after checking the `[X] dir` and clicking `[Ok]` the map is there again ...

**Comment by Billy-Wallace** [comment]
Just to say, I like what you are doing here and it's a great additional feature for the already fantastic c:geo (Thanks guys). To make it more in line with Settings -> Directory with offline maps and Map Themes Directory, why not include a "Map Routes Directory" so that c:geo has a default location to scan when loading a new route.

**Comment by moving-bits** [comment]
@Billy-Wallace 
c:geo internally stores the dir where you have your GPX files stored in, and already uses this for tracks and individual routes as well.
